lib/internal/util/inspect_colors.js

```diff
@@ -0,0 +1,498 @@
+        currentPart = '';
+        inValue = false;
+        continue;
+      }
+    } else if (c === ':') {
+      currentKey = StringPrototypeTrim(currentPart);
+      currentPart = '';
+      inValue = true;
+      continue;
+    }
+    currentPart += c;
+  }
+  if (inValue && parenthesesDepth === 0) {
+    const value = StringPrototypeTrim(currentPart);
+    if (value !== '') {
+      ArrayPrototypePush(rawEntries, [currentKey, value]);
+    }
+    currentKey = null;
+    currentPart = '';
+  }
+
+  for (let i = 0; i < rawEntries.length; ++i) {
+    const { 0: key, 1: value } = rawEntries[i];
+    if (key === 'background-color') {
+      if (value != null) {
+        css.backgroundColor = value;
+      }
+    } else if (key === 'color') {
+      if (value != null) {
+        css.color = value;
+      }
+    } else if (key === 'font-weight') {
+      if (value === 'bold') {
+        css.fontWeight = value;
+      }
+    } else if (key === 'font-style') {
+      if (
+        ArrayPrototypeIncludes(['italic', 'oblique', 'oblique 14deg'], value)
+      ) {
+        css.fontStyle = 'italic';
+      }
+    } else if (key === 'text-decoration-line') {
+      css.textDecorationLine = [];
+      const lineTypes = StringPrototypeSplit(value, SPACE_PATTERN);
+      for (let i = 0; i < lineTypes.length; ++i) {
+        const lineType = lineTypes[i];
+        if (
+          ArrayPrototypeIncludes(
+            ['line-through', 'overline', 'underline'],
+            lineType,
+          )
+        ) {
+          ArrayPrototypePush(css.textDecorationLine, lineType);
+        }
+      }
+    } else if (key === 'text-decoration-color') {
+      const color = parseCssColor(value);
+      if (color != null) {
+        css.textDecorationColor = color;
+      }
+    } else if (key === 'text-decoration') {
+      css.textDecorationColor = null;
+      css.textDecorationLine = [];
+      const args = StringPrototypeSplit(value, SPACE_PATTERN);
+      for (let i = 0; i < args.length; ++i) {
+        const arg = args[i];
+        const maybeColor = parseCssColor(arg);
+        if (maybeColor != null) {
+          css.textDecorationColor = maybeColor;
+        } else if (
+          ArrayPrototypeIncludes(['line-through', 'overline', 'underline'], arg)
+        ) {
+          ArrayPrototypePush(css.textDecorationLine, arg);
+        }
+      }
+    }
+  }
+
+  return css;
+}
+
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+// https://github.com/denoland/deno/blob/ece2a3de5b19588160634452638aa656218853c5/ext/console/01_console.js#L2928
+function colorEquals(color1, color2) {
+  return color1?.[0] === color2?.[0] && color1?.[1] === color2?.[1] &&
+    color1?.[2] === color2?.[2];
+}
+
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+// https://github.com/denoland/deno/blob/ece2a3de5b19588160634452638aa656218853c5/ext/console/01_console.js#L2933
+function cssToAnsi(css, prevCss = null) {
```

Comment:
I think this should be using `utile.styleText` instead of reimplementing it. Unless there's a good reason not to use it? If so we should add a comment.

Reply:
util.styleText() uses util.inspect.colors.

https://github.com/nodejs/node/pull/49205#discussion_r1593255270

> after I tried digging into it more I don't think that using util.inspect.colors is a good idea for console.log() %c.
> 
> 1. util.inspect.colors is mutable. this can be good and bad. on one hand it lets users customize what "red" means. on the other had, it lets users change what "red" means but NOT what #ff0000 means. changing what "red" means is something not even browsers let you do (to my knowledge)!
> 2. I don't think it would be a good idea to pull out just the magic ~15 or so color names and have those reference some object defined in a different file which every _other_ color (like "salmon") uses color->hex translation. it creates two files to dig through to follow instead of sticking everything in one spot.
> 
> ex for number 1:
> 
> ```js
> util.inspect.colors.red = util.inspect.colors.green
> console.log("%cI am red!", "color:red") // what color am I?
> console.log("%cI am red!", "color:#ff0000") // what color am I?
> ```
> 
> should these print the same color? or different? what about salmon:
> 
> ```js
> // should attaching a new property to util.inspect.colors
> // add a new CSS->ANSI color translation?
> util.inspect.colors.salmon = util.inspect.colors.green
> console.log("%cI am red!", "color:salmon") // what color am I?
> console.log("%cI am red!", "color:#fa8072") // what color am I?
> ```
> 
> util.inspect.colors is mutable by the user which is a whole can of worms. do you make a copy of it? if so, why not just define the literals in the file; why reference util.inspect.colors at all for ~15 color codes -- its almost easier to just define the object in the same file lol.
> 
> for number 2 to elaborate a bit... i also don't think its a good idea to deviate too far from deno's original code so that its easier to copy-paste anyu changes in the intermediate future and say "thanks deno devs!" instead of grafting any updates onto custom node.js util.inspect.colors in-another-file code.
> 
> relevant code bits
> 
> [jcbhmr/node@`c130398`/lib/internal/util/inspect_colors.js#L20](https://github.com/jcbhmr/node/blob/c130398eb5e7f6334f295956abc5a4b950c79af6/lib/internal/util/inspect_colors.js#L20)
> 
> [jcbhmr/node@`c130398`/lib/internal/util/inspect_colors.js#L373](https://github.com/jcbhmr/node/blob/c130398eb5e7f6334f295956abc5a4b950c79af6/lib/internal/util/inspect_colors.js#L373)
> 
> and btw I misinterpreted that original git blame I commented saying "oh deno updated their code"; that was just deno adding some formatArgs() to mimic nodejs util.format().
> 
> so tldr I don't think deviating from deno's code (which is actually totally unchanged in the color section that we are discussing; they only modified the other formatting-related code, not the CSS color %c magic) is a good idea.

tldr of the quote:
1. util.styleText uses util.inspect.colors. util.inspect.colors is _mutable_. this is bad because some colors (red, green, blue) are editable while others (salmon, deepskyblue, `#ff0000`) are not editable.
2. I don't think its a good idea to split color CSS->ANSI processing over two different modules/files/domains for maintenance and readability

Reply:
FWIW I don't find the arguments for not using `util.inspect.colors` very compelling: I don't think we should be worry about user mutability (on the contrary, it can be a plus from an accessibility perspective), and from a maintenance perspective having one list of colors is better than two.

Reply:
> it can be a plus from an accessibility perspective

I've seen this mentioned before about how "mutable colors are good" and I want to refute this.

1. It's unreasonable to expect any end-user who is colorblind or has a vision impairment to use `node --import 'data:text/javascript,import { inspect } from "node:util"; inspect.colors.red = inspect.colors.gray' node_modules/myapp/bin/cli.js`. Or even `NODE_OPTIONS='--import "data:text/javascript,..."'`. That's just plain _not portable_ to the numerous other CLI applications like `make`, `ripgrep`, `cargo`, etc.; it only works on Node.js apps that are using the builtin `inspect.colors` mapping and don't pull in https://www.npmjs.com/package/ansi-colors or similar npm packages which define their own copies. Editing the "which ANSI code is for `red`?" information _as an end user_ is completely unfeasible and so high effort/reward to be overlooked in favor of broader terminal settings or system visual settings.

2. Not all apps are Node.js apps where this would apply. Think of GNU coreutils. Instead, a user will probably use a terminal emulator which custom-interprets ANSI color codes and remaps them at the terminal level. The number one reason you would want to do this is color blindness. Apps often have ways to force no-color text per-app either with flags or `NO_COLOR` or `TERM=dumb` or `mycmd | cat` or similar. Another more global way is disabling ANSI color codes in your terminal emulator or forcing a minimum contrast like iTerm does https://iterm2.com/documentation-preferences-profiles-colors.html#:~:text=faint%22%20text%20is.-,Minimum%20Contrast,-Sometimes%20text%20in which when maxed effectively disables ANSI color codes. In Windows Terminal you use https://learn.microsoft.com/en-us/windows/terminal/customize-settings/color-schemes#creating-your-own-color-scheme to set a custom alias for the common (non hex code) colors to suit your needs. Ubuntu terminal can redefine ANSI colors https://itsfoss.com/change-terminal-color-ubuntu/ as well.

3. Most apps _want to be sure_ that they are using _that color_. From a developer perspective having something that's _mutable_ depending on whether you do `const red = inspect.colors.red[0]` or `const getRed = () => inspect.colors.red[0]` is a bit baffling. You generally **don't want to encourage mutating such a critical core color map**. **Especially since it will become out of sync with other npm packages which maintain their own ANSI map like https://www.npmjs.com/package/ansi** and now you have two different definitions for the color `green`. Unlikely (I hope) but not improbable.

4. inspect.colors mutation is **actually not used** for accessibility. Or much at all for that matter. From a GitHub search https://github.com/search?q=%2Finspect%5C.colors%5B%5C%5B%5C.%5D%5Cw%2B%5C%5D%3F%5Cs*%3D%2F&type=code for `/inspect\.colors[\[\.]\w+\]?\s*=/` there are only _eight results_, only one of which actually overwrites a color to `0, 0, 0` for a test. Sourcegraph search for the same https://sourcegraph.com/search?q=context:global+/inspect%5C.colors%5B%5C%5B%5C.%5D%5Cw%2B%5C%5D%3F%5Cs*%3D/&patternType=keyword&sm=0 yields 37 results across 6 files. Same thing; defining new colors and DefinitelyTyped test. The preferred way for applications to allow color customization seems to be through environment variables like `LS_COLORS` & `MYCLI_THEME=catppuccin` or CLI flags like `--theme=light`. Or `NO_COLOR=1` to disable output https://no-color.org/

5. Node.js recommends using the higher-level `inspect.styles` to choose custom colors instead of mutating the global `inspect.colors` map. https://nodejs.org/api/util.html#customizing-utilinspect-colors 😅

6. Browsers are up there with most widely used software ever and **do not allow developers to customize what `color: red; background-color: blue` means**. Instead they provide the user with tools to view the page their own way _at the viewer level_ similar to how a terminal emulator provides knobs to customize terminal colors. Like turning off JavaScript https://www.impressivewebs.com/how-to-disable-javascript-in-almost-any-browser/ or disabling CSS stylesheets https://smallbusiness.chron.com/disable-firefox-stylesheets-49055.html or even (albiet with extensions) overlaying custom styles https://darkreader.org/ or my favorite: https://chromewebstore.google.com/detail/colorblind-dalton-for-goo/afcafnelafcgjinkaeohkalmfececool

While I agree that making `util.inspect.colors` immutable now is never going to happen (`inspect.colors.MYCOLOR=...; inspect.style.number="MYCOLOR"` since `inspect.style.*` uses `inspect.color.*` properties -- is still needed), I don't think actively citing "mutating `util.inspect.colors` is an opportunity for accessibility since it's a centralized map that can be customized" is a good idea. There are other better & higher-level ways to improve accessibility without messing with color name/value pairs.

---

> and from a maintenance perspective having one list of colors is better than two.

<sup>I don't think the ANSI core color codes are going to be extended or added to any time soon lol 😉</sup>

The main maintenance I can imagine is copy-pasting code from https://github.com/denoland/deno/blob/main/ext/console/01_console.js#L2449-L2931 which is a pretty easy thing to do right now; it's all the same code (with some primordial tweaks)! If you start changing more stuff, it becomes more of a hassle. 🤷‍♀️

---

If this code:

```js
if (css.backgroundColor == null) {
  ansi += '\x1b[49m';
} else if (css.backgroundColor === 'black') {
  ansi += '\x1b[40m';
} else if (css.backgroundColor === 'red') {
  ansi += '\x1b[41m';
} else if (css.backgroundColor === 'green') {
  ansi += '\x1b[42m';
} else if (css.backgroundColor === 'yellow') {
  ansi += '\x1b[43m';
} else if (css.backgroundColor === 'blue') {
  ansi += '\x1b[44m';
} else if (css.backgroundColor === 'magenta') {
  ansi += '\x1b[45m';
} else if (css.backgroundColor === 'cyan') {
  ansi += '\x1b[46m';
} else if (css.backgroundColor === 'white') {
  ansi += '\x1b[47m';
} else if (ArrayIsArray(css.backgroundColor)) {
  const { 0: r, 1: g, 2: b } = css.backgroundColor;
  ansi += `\x1b[48;2;${r};${g};${b}m`;
} else {
  const parsed = parseCssColor(css.backgroundColor);
  if (parsed !== null) {
    const { 0: r, 1: g, 2: b } = parsed;
    ansi += `\x1b[48;2;${r};${g};${b}m`;
  } else {
    ansi += '\x1b[49m';
  }
}
```

were changed to this:

```js
if (css.backgroundColor == null) {
  ansi += '\x1b[49m';
} else if (css.backgroundColor === 'black') {
  ansi += `\x1b[${inspect.colors.bgBlack[0]}m`;
} else if (css.backgroundColor === 'red') {
  ansi += `\x1b[${inspect.colors.bgRed[0]}m`;
} else if (css.backgroundColor === 'green') {
  ansi += `\x1b[${inspect.colors.bgGreen[0]}m`;
} else if (css.backgroundColor === 'yellow') {
  ansi += `\x1b[${inspect.colors.bgYellow[0]}m`;
...
```

or

```js
if (css.backgroundColor == null) {
  ansi += '\x1b[49m';
} else if (ArrayPrototypeIncludes(['black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white'], css.backgroundColor)) {
  const upperBg = stringPrototypeToUpperCase(css.backgroundColor[0]) + StringPrototypeSlice(css.backgroundColor, 1);
  ansi += `\x1b[${inspect.color[`bg${upperBg}`][0]}m`;
} else {
  const parsed = parseCssColor(css.backgroundColor);
  if (parsed !== null) {
    const { 0: r, 1: g, 2: b } = parsed;
    ansi += `\x1b[48;2;${r};${g};${b}m`;
  } else {
    ansi += '\x1b[49m';
  }
}
```

I don't think that unifies things the way one would hope. There's still the massive `Map<string, string>` holding all the `salmon`=>`#FF8C69` color names/values; should that be stuffed into `inspect.colors` too? I don't know. The crux is that the `inspect.colors` map doesn't include all the colors that CSS colors need, while at the same time also being mutable. Where does the line between internal color map and user-mutable `inspect.colors` map get drawn?

In the way the code is above ☝ the core ANSI colors (`background:red`, `color:blue`) are user-mutable vs but CSS colors (salmon, deepskyblue) or hex colors (`#223344`) aren't. ...which is weird and doesn't feed right to me. I'd prefer uniformity and a guarantee of "I say `color:red`, I get _red_ same as if I said `color:#ff0000`; the terminal user can `NO_COLOR=1` or `MYAPP_THEME=colorblind2` or edit their terminal settings if they need to." as a developer. 

If you want to though I guess I can rewrite those 20 if statement lines. I need some feedback on some questions as to how to rewrite the code to use inspect.colors though:

- Which ANSI code should this code use? `inspect.colors.salmon = inspect.colors.red; console.log("%cfoo", "color:salmon")`
- Which ANSI code should this code use? `inspect.colors.red = inspect.colors.green; console.log("%cfoo, "color:#ff0000")`
- Which ANSI code should this code use? Should it throw? `delete inspect.colors.red; console.log("%cfoo", "color:red")`
- Which ANSI code should this code use? Should it throw? `inspect.colors.red = "hi"; console.log("%cfoo", "color:red")`
- What's the current behaviour for `util.format()` when a `inspect.colors` is an invalid type?
- Should `console.log()` _ever_ be able to throw an error?
- Which ANSI code should this code use? `inspect.colors["#111111"]=inspect.colors.red; console.log("%cfoo", "color:#111111")`

Reply:
Wow! That is detailed! IMO your current implementation works, and it can always be changed later.

In the long term, what if (and this is only a theoretical idea), in a seperate PR (after this lands), `util.colors`  was extended to include the colors listed here, and this can pull from both hexes *and* the `util.colors` map?

I'm happy to implement that once this lands.

Reply:
> 4\. In the Real World™ inspect.colors mutation doesn't seem to be used for accessibility. Or much at all for that matter. From a GitHub search https://github.com/search?q=%2Finspect%5C.colors%5B%5C%5B%5C.%5D%5Cw%2B%5C%5D%3F%5Cs*%3D%2F&type=code for `/inspect\.colors[\[\.]\w+\]?\s*=/` there are only _eight results_

That seems to further confirm that mutability is a non-problem 🤷‍♂️  

> 3\. Most apps _want to be sure_ that they are using _that color_

The user is the one reading the text in the end, so it's good that they can control/tweak it, and how the app author feels about that seems irrelevant to me.


> should that be stuffed into `inspect.colors` too?

Yes, like I said, 1 list is better than 2.



> * Should `console.log()` _ever_ be able to throw an error?

In an ideal world, no (note that it's already not the case, it's possible that calling `console.log` result in an error when some internals have been mutated, but IMO we should not add more cases where that happens).



> * Which ANSI code should this code use? `inspect.colors.salmon = inspect.colors.red; console.log("%cfoo", "color:salmon")`
> * Which ANSI code should this code use? `inspect.colors.red = inspect.colors.green; console.log("%cfoo, "color:#ff0000")`
> * Which ANSI code should this code use? Should it throw? `delete inspect.colors.red; console.log("%cfoo", "color:red")`
> * Which ANSI code should this code use? Should it throw? `inspect.colors.red = null; console.log("%cfoo", "color:red")`
> * Which ANSI code should this code use? `inspect.colors["#111111"]=inspect.colors.red; console.log("%cfoo", "color:#111111")`

I think it's fine to assume `inspect.colors` won't be mutated, and so it's fine to have undefined behavior.

> * Which ANSI code should this code use? `console.log("%cfoo", "color:bgBlue")`

Given it's not valid CSS, I think it's fine if it has undefined behavior (e.g. if it behaves similar to `background-color:blue`).

> * Which ANSI code should this code use? `console.log("%cfoo", "color:hasOwnProperty")`

`hasOwnProperty` is not a property of `inspect.colors`.

---

This is just my opinion, feel free to disagree, I'm not a user of that feature anyway, and my comments are non-blocking.

Reply:
While I like the idea of one list rather than two, my only concern is that there is no background vs foreground on the util colors list (bgBlue etc, but that's not the same).

I think that this PR looks okay for now, and once it lands a more long term change to util colors and this might be a valuable thing to discuss.

Then again, I'm a triage member, so I don't really have an opinion on this.

Reply:
@aduh95 my 6-point rant wasn't really about how it being mutable was bad but rather that "mutable colors are good _because accessibility_" is a bit disingenuous. Encouraging the average colorblind end-user to "just `--import` preload and change inspect.colors" or the average dev to "let the user manually mutate inspect.colors to change theme" is a bad recommendation. I do think the overall goal of consolidating a color list in one spot is a good idea.

> I think it's fine to assume inspect.colors won't be mutated, and so it's fine to have undefined behavior.

Thanks for clarifying; I was confused earlier about encouraging mutating colors for accessibility reasons. 👍

Reply:
Maybe, but that's an issue for a different PR. Let's focus on this one landing first